File: sample_question/custom_validator/sample-validator.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	exitAC  = 0
	exitWA  = 1
	exitTLE = 2
	exitRE  = 3
	exitPE  = 4
)

var reader *bufio.Reader

func finish(exitcode int, message string) {
	if message != "" {
		fmt.Fprintln(os.Stderr, message)
	}
	os.Exit(exitcode)
}

func readLine() (string, bool) {
	line, err := reader.ReadString('\n')
	if err != nil && line == "" {
		return "", false
	}
	return strings.TrimSpace(line), true
}

func parseInteger(text string) (int64, bool) {
	value, err := strconv.ParseInt(strings.TrimSpace(text), 10, 64)
	return value, err == nil
}

func envNumber(name string) string {
	value, ok := os.LookupEnv(name)
	if !ok {
		return "null"
	}
	return value
}

func printLimits() {
	userLanguage := "null"
	if value, ok := os.LookupEnv("USER_LANGUAGE"); ok {
		userLanguage = strconv.Quote(value)
	}
	perLanguageLimits := "null"
	if value, ok := os.LookupEnv("PER_LANGUAGE_LIMITS"); ok {
		perLanguageLimits = value
	}

	fmt.Fprintf(os.Stderr,
		"{\"problem_time_limit\": %s, \"problem_output_limit\": %s, \"problem_memory_limit\": %s, "+
			"\"problem_pid_limit\": %s, \"user_language\": %s, \"per_language_limits\": %s}\n",
		envNumber("PROBLEM_TIME_LIMIT"),
		envNumber("PROBLEM_OUTPUT_LIMIT"),
		envNumber("PROBLEM_MEMORY_LIMIT"),
		envNumber("PROBLEM_PID_LIMIT"),
		userLanguage,
		perLanguageLimits,
	)
}

func readSecretValue() int64 {
	line, ok := readLine()
	if !ok {
		finish(exitRE, "Invalid secret data")
	}
	value, ok := parseInteger(line)
	if !ok {
		finish(exitRE, "Invalid secret data")
	}
	return value
}

func main() {
	reader = bufio.NewReader(os.Stdin)

	maxValue := readSecretValue()
	maxTries := readSecretValue()
	secret := readSecretValue()

	printLimits()

	fmt.Println(maxValue)

	for iteration := int64(0); ; iteration++ {
		user, ok := readLine()
		if !ok {
			break
		}

		if strings.HasPrefix(user, "!") {
			answer, ok := parseInteger(user[1:])
			if !ok {
				finish(exitRE, "Invalid data")
			}
			if answer == secret {
				finish(exitAC, "")
			}
			finish(exitWA, fmt.Sprintf("!%d", secret))
		}

		if iteration > maxTries {
			finish(exitTLE, "Exceeded number of allowed iterations")
		}
		value, ok := parseInteger(user)
		if !ok {
			finish(exitRE, "Invalid data")
		}
		if secret < value {
			fmt.Println("<")
		} else {
			fmt.Println(">=")
		}
	}

	finish(exitRE, "Unexpected flow")
}

File: sample_question/custom_validator/sample-validator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum { EXIT_AC = 0, EXIT_WA = 1, EXIT_TLE = 2, EXIT_RE = 3, EXIT_PE = 4 };

static void finish(int exitcode, const char *message) {
    if (message != NULL) {
        fprintf(stderr, "%s\n", message);
        fflush(stderr);
    }
    exit(exitcode);
}

static int read_line(char *buffer, size_t size) {
    if (fgets(buffer, (int)size, stdin) == NULL) {
        return 0;
    }
    buffer[strcspn(buffer, "\r\n")] = '\0';
    return 1;
}

static int parse_long(const char *text, long *value) {
    char *end = NULL;
    if (*text == '\0') {
        return 0;
    }
    *value = strtol(text, &end, 10);
    while (*end == ' ' || *end == '\t') {
        end++;
    }
    return *end == '\0';
}

static void print_limits(void) {
    static const char *names[] = {
        "PROBLEM_TIME_LIMIT",
        "PROBLEM_OUTPUT_LIMIT",
        "PROBLEM_MEMORY_LIMIT",
        "PROBLEM_PID_LIMIT",
    };
    static const char *keys[] = {
        "problem_time_limit",
        "problem_output_limit",
        "problem_memory_limit",
        "problem_pid_limit",
    };
    const char *user_language = getenv("USER_LANGUAGE");
    const char *per_language_limits = getenv("PER_LANGUAGE_LIMITS");

    fputc('{', stderr);
    for (int index = 0; index < 4; index++) {
        const char *value = getenv(names[index]);
        fprintf(stderr, "\"%s\": %s, ", keys[index], value != NULL ? value : "null");
    }
    if (user_language != NULL) {
        fprintf(stderr, "\"user_language\": \"%s\", ", user_language);
    } else {
        fprintf(stderr, "\"user_language\": null, ");
    }
    fprintf(stderr, "\"per_language_limits\": %s}\n",
            per_language_limits != NULL ? per_language_limits : "null");
    fflush(stderr);
}

int main(void) {
    char line[4096];
    long max_value = 0;
    long max_tries = 0;
    long secret = 0;

    if (!read_line(line, sizeof line) || !parse_long(line, &max_value)) {
        finish(EXIT_RE, "Invalid secret data");
    }
    if (!read_line(line, sizeof line) || !parse_long(line, &max_tries)) {
        finish(EXIT_RE, "Invalid secret data");
    }
    if (!read_line(line, sizeof line) || !parse_long(line, &secret)) {
        finish(EXIT_RE, "Invalid secret data");
    }

    print_limits();

    printf("%ld\n", max_value);
    fflush(stdout);

    for (long iteration = 0; read_line(line, sizeof line); iteration++) {
        long value = 0;
        if (line[0] == '!') {
            if (!parse_long(line + 1, &value)) {
                finish(EXIT_RE, "Invalid data");
            }
            if (value == secret) {
                finish(EXIT_AC, NULL);
            }
            char message[64];
            snprintf(message, sizeof message, "!%ld", secret);
            finish(EXIT_WA, message);
        }

        if (iteration > max_tries) {
            finish(EXIT_TLE, "Exceeded number of allowed iterations");
        }
        if (!parse_long(line, &value)) {
            finish(EXIT_RE, "Invalid data");
        }
        printf("%s\n", secret < value ? "<" : ">=");
        fflush(stdout);
    }

    finish(EXIT_RE, "Unexpected flow");
    return EXIT_RE;
}
